templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml

```diff
@@ -20,89 +20,121 @@ contents:
-                    sed -i "s/^search \(.*\)/search {{.DNS.Spec.BaseDomain}} \1/" /etc/resolv.tmp
-                    # Remove duplicate cluster domain entries
-                    sed -i "s/\(search {{.DNS.Spec.BaseDomain}}.*\) {{.DNS.Spec.BaseDomain}}\( .*\|$\)/\1\2/" /etc/resolv.tmp
+                    >&2 echo "NM resolv-prepender: Prepending 'nameserver $NAMESERVER_IP' to /etc/resolv.conf (other nameservers from /var/run/NetworkManager/resolv.conf)"
+                    RESOLVCONFTMPPATH="$(mktemp)"
+                    sed -e "/Generated by/c# Generated by KNI resolv prepender NM dispatcher script" \
+                        /var/run/NetworkManager/resolv.conf > "${RESOLVCONFTMPPATH}"
+                    sed -i "0,/^nameserver.*/s//nameserver $NAMESERVER_IP\n\0/" "${RESOLVCONFTMPPATH}"
+                    if ! grep -q search "${RESOLVCONFTMPPATH}"; then
+                        # Make sure we have a search entry
+                        echo "search {{.DNS.Spec.BaseDomain}}" >> "${RESOLVCONFTMPPATH}"
+                    else
+                        # Make sure cluster domain is first in the search list
+                        sed -i "s/^search \(.*\)/search {{.DNS.Spec.BaseDomain}} \1/" "${RESOLVCONFTMPPATH}"
+                        # Remove duplicate cluster domain entries
+                        sed -i "s/\(search {{.DNS.Spec.BaseDomain}}.*\) {{.DNS.Spec.BaseDomain}}\( .*\|$\)/\1\2/" "${RESOLVCONFTMPPATH}"
+                    fi
+                    # Only leave the first 3 nameservers in /etc/resolv.conf
+                    sed -i ':a $!{N; ba}; s/\(^\|\n\)nameserver/\n# nameserver/4g' "${RESOLVCONFTMPPATH}"
+                    mv -f "${RESOLVCONFTMPPATH}" /etc/resolv.conf
+                    # Workaround for bz 1929160. Reload NetworkManager to force it to
+                    # re-run the lookup of the hostname now that we know we have DNS
+                    # servers configured correctly in resolv.conf.
+                    nmcli general reload dns-rc
                 fi
-                # Only leave the first 3 nameservers in /etc/resolv.conf
-                sed -i ':a $!{N; ba}; s/\(^\|\n\)nameserver/\n# nameserver/4g' /etc/resolv.tmp
-                mv -f /etc/resolv.tmp /etc/resolv.conf
-                # Workaround for bz 1929160. Reload NetworkManager to force it to
-                # re-run the lookup of the hostname now that we know we have DNS
-                # servers configured correctly in resolv.conf.
-                nmcli general reload dns-rc
             fi
-        fi
-        ;;
-        *)
-        ;;
-    esac
+            ;;
+            *)
+            ;;
+        esac
+    }
+
+    # Pull container image outside of the timeout in a background process to workaround possible
+    # image corruption [1]. As long as the overall script duration is below NM dispatcher timeout, this pull will continue
+    # in the background even after this script concludes [1].
```

Comment:
hrrm so if the dispatcher script ends, this will be orphaned, detached from parent and inherited by init or whatever the session manager is, so that it continues to run in the background without NM ever being aware of it. Perhaps adding this extra minor detail in the comment might be useful, but this is really a nit, up to you.

Reply:
`If the script ends by itself before the NM timeout and before this pull is done, this pull will be orphaned from the parent and continue running without NM being aware of it`